src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs

```diff
@@ -476,33 +476,45 @@ internal Diagnostic WithProgrammaticSuppression(ProgrammaticSuppressionInfo prog
         /// </summary>
         internal bool HasIntersectingLocation(SyntaxTree tree, TextSpan? filterSpanWithinTree = null)
         {
-            var locations = this.GetDiagnosticLocationsWithinTree(tree);
+            if (AdditionalLocations is null || AdditionalLocations.Count == 0)
+            {
+                // Fast path to avoid allocations.
+                // Test scenario: https://github.com/dotnet/roslyn/issues/26778
+                return Location.SourceTree == tree
+                    && isLocationWithinSpan(Location, filterSpanWithinTree);
+            }
 
-            foreach (var location in locations)
+            foreach (var location in getDiagnosticLocationsWithinTree(Location, AdditionalLocations, tree))
             {
-                if (!filterSpanWithinTree.HasValue || filterSpanWithinTree.Value.IntersectsWith(location.SourceSpan))
+                if (isLocationWithinSpan(location, filterSpanWithinTree))
                 {
                     return true;
                 }
             }
 
             return false;
-        }
 
-        private IEnumerable<Location> GetDiagnosticLocationsWithinTree(SyntaxTree tree)
-        {
-            if (this.Location.SourceTree == tree)
+            // Local functions
+            static bool isLocationWithinSpan(Location location, TextSpan? filterSpan)
             {
-                yield return this.Location;
+                return !filterSpan.HasValue || filterSpan.GetValueOrDefault().IntersectsWith(location.SourceSpan);
             }
 
-            if (this.AdditionalLocations != null)
+            static IEnumerable<Location> getDiagnosticLocationsWithinTree(Location location, IReadOnlyList<Location> additionalLocations, SyntaxTree tree)
             {
-                foreach (var additionalLocation in this.AdditionalLocations)
+                if (location.SourceTree == tree)
+                {
+                    yield return location;
+                }
+
+                if (additionalLocations is object && additionalLocations.Count > 0)
```

Comment:
>if (additionalLocations is object && additionalLocations.Count > 0) [](start = 16, length = 67)

These cases are already handled in the caller.

Reply:
I'll change it to an assertion.